### DISCUSSION TITLE
Dealing with TabControls

### DISCUSSION THREAD

**Discussion by B-B-Christian** [AvaloniaCommunity/Prism.Avalonia#88] [Q&A] [answered]
First of all, i wanted to thank you for your work on this project.

I've been wondering what would be the correct way to handle TabControls in Prism.Avalonia.
So far my approach has been cobbled together from old Prism.WPF videos and StackOverflow questions, as i wasn't able to find informations about Prism.Avalonia on that regard.
After metaphorically glueing stuff together i ended up having that TabControl be in its own region, making a region adapter, mapping TabControl with the adapter and registering views and viewmodels for the tabs in that region.
So far so good, in that at least it works in managing the creation and use of those tabs.
Changing tabs doesn't trigger the OnNavigatedTo method though.
I did read some old discussions about past versions of Prism.WPF where people mentioned using a behaviour triggered on the Loaded event to call a command as a workaround - in Avalonia's case i guess we could use the Opened event - but i have to wonder if it is indeed the correct way to go at it.

**Comment by DamianSuess** [comment]
Thank you @B-B-Christian for reaching out and for your support.

If you don't mind, could you reference the StackOverflow and videos that you're referencing? I just want to make sure that I have a full grasp of the intent of what you're trying to perform. No doubt, a sample project could be created for working with Tabs

**Comment by B-B-Christian** [comment]
Gotta rely on memory since i've been working on this project for some time, but for the adapter part, one of my main sources of info ended up being this youtube video (granted, given i hadn't used either WPF, Avalonia or Prism before, it took some searching just to understand and learn that you do have to make a separate region to make a TabControl work to begin with).
https://www.youtube.com/watch?v=BXeIZgssP0A
WPF of course, but as far as latest Avalonia and Prism versions go, the code works (i remember i had to change a bunch of stuff during the Avalonia 11 preview phase due to some issues, but well, it was the preview versions after all).
Thus ending up applying a region to the TabControl, registering types for views and viewmodels of the tabs, registering views for navigation in that region (made that via a separate module, just to keep things orderly), making a TabControlAdapter and registering the mapping for TabControl to that adapter. 
I did make things a bit more complex in my case by customizing the tabitems and adding a back button to the UserControl that contains the TabControl, but those aren't strictly Prism-related, so i'm skipping those for the sake of conciseness.
In terms of a workaround for a tab selection event i had found several stuff, ranging from implementing IActiveAware interfaces on viewmodels and using the IsActive property (but those StackOverflow questions were very outdated and about Xamarin.Prism, and i also ended up finding this other question
https://stackoverflow.com/questions/50102026/how-to-run-a-method-every-time-a-tabitem-is-selected-in-an-mvvm-application-usi 
On that regard i noticed i had misread an older avalonia issue (didn't notice it was about a window) and so thought there wasn't a "Loaded" event for views, as i wrongly mentioned above.
So by using Avalonia.Xaml.Interactivity and Avalonia.Xaml.Interactions, i can have something like this in the tab views
```
<i:Interaction.Behaviors>
    <ia:EventTriggerBehavior SourceObject="{Binding #NameOfTabUserControl}" EventName="Loaded">
      <ia:InvokeCommandAction Command="{Binding OnLoadedCommand}" />
    </ia:EventTriggerBehavior>
  </i:Interaction.Behaviors>
```
With the command calling a method from viewmodel.
While it doesn't really take the place of the OnNavigatedTo method, it at least allows to have a sort of OnTabSelected method (with some caveats, as it also triggers when getting back from subviews, so some checks may be required in order to make stuff work properly).

What i was wondering was mainly the correctness of such an approach, though, both for the adapter part and this other workaround.
As in, it's that how we're supposed to work with tabs, or i'm instead making a mess of something that's supposed to be more, ehr, refined?

**Reply by DamianSuess** [reply, marked as answer]
I see what you mean, @B-B-Christian. Though you do bring up an interesting point, thus far, my implementations of `TabControl` in Avalonia thus far have been pretty simple, not relying on the need for navigation strategies - _except for my MAUI/Xamarin.Forms apps_.

While awaiting your response, I was thinking along the same lines as your code reference, [Avalonia.Xaml.Behaviors](https://github.com/AvaloniaUI/Avalonia.Xaml.Behaviors). Based on your inquiry and samples, the [Prism.Avalonia Outlookish](https://github.com/DamianSuess/Learn.PrismAvaloniaOutlookish) project will be getting a TabControl update with something similar in mind.

### Check out the PR:

This pull request is an example of using the TabControlAdapter based on the sample video. This, however, does not provide the IActiveAware linkages.

* https://github.com/DamianSuess/Learn.PrismAvaloniaOutlookish/pull/10
* https://github.com/DamianSuess/Learn.PrismAvaloniaOutlookish/pull/11

You're on the right track for sure.

For another example of using Prism's `RegionAdapterBase`, check out Brian Lagunas' video on **Custom Region Adapter**
* https://youtu.be/-DvtGorOfD8?t=192
* https://app.pluralsight.com/library/courses/prism-mastering-tabcontrol/table-of-contents

**Reply by DamianSuess** [reply]
Example Code for Avalonia v0.10.x

In short, this allows for double tapping of a ListBox's item to invoke the Command `CmdCopyText`

```xml
  <UserControl.Resources>
    <ResourceDictionary>
      <DataTemplate x:Key="GridDeviceInfo" DataType="models:DeviceInfo">
        <Panel Background="Transparent">
          <i:Interaction.Behaviors>
            <ia:EventTriggerBehavior EventName="DoubleTapped">
              <!-- Must set the parent, otherwise it will look for the 'Command' in the template's DataType -->
              <ia:InvokeCommandAction Command="{Binding $parent.((vm:AccessViewModel)DataContext).CmdCopyText}"
                                      CommandParameter="{Binding Id}" />
              <!-- In some cases you may have to specify the `[Grid]`
              <ia:InvokeCommandAction Command="{Binding $parent[Grid].((vm:AccessViewModel)DataContext).CmdCopyText}"
                                      CommandParameter="{Binding Id}" />
              -->
            </ia:EventTriggerBehavior>
          </i:Interaction.Behaviors>

          <Grid ColumnDefinitions="Auto,8,*" RowDefinitions="Auto,Auto">
           ...
  </UserControl.Resources>
  ...
    <ListBox ItemTemplate="{StaticResource GridDeviceInfo}"
             Items="{Binding Devices}"
             SelectedItem="{Binding DeviceSelected, Mode=TwoWay}" />
```
